### ISSUE TITLE
ReferenceError: self is not defined ( Mocha+JSdom )

### ISSUE THREAD

**Issue by dlin-me** [webpack-contrib/style-loader#177]
## Version 

0.13.2

## Request

Bug

## Current behavior

Got the following error when running test cases with mocha+jsdom. **Same test run fine with version 0.13.1**

```
ReferenceError: self is not defined
    at eval (Users/dlin/Projects/insiight/onboarding/node_modules/style-loader/addStyles.js:14:30)
    at eval (Users/dlin/Projects/insiight/onboarding/node_modules/style-loader/addStyles.js:9:47)
    at module.exports (Users/dlin/Projects/insiight/onboarding/node_modules/style-loader/addStyles.js:31:68)
```

**Comment by ekulabuhov** [comment]
Hi! Do you mind providing a bare bone repro of your issue?

**Comment by frodare** [comment]
This patch broke our testing build.  After looking into the problem we were able to fix it by defining a global self variable in our jsdom setup like this:

```
var jsdom = require('jsdom');
global.document = jsdom.jsdom('<!DOCTYPE html><head></head><html><body></body></html>');
global.window = document.defaultView;
global.window.document = global.document;
// fix
global.self = global.window;
```

**Comment by ekulabuhov** [comment]
This issue was caused by https://github.com/webpack-contrib/style-loader/commit/9dc45a6dee16553a64071031d18feb1562df86a2

The problem here is that `self` is only defined in Browsers and Web Workers. Unit tests running in NodeJs environment won't have access to `self`.

**Comment by FlyingDR** [comment]
Same issue breaks attempts to use `style-loader` to build Firefox add-ons (at least SDK-based) since `self` here contains no `navigator` property and it results into `TypeError: self.navigator is undefined` error

**Comment by ekulabuhov** [comment]
`style-loader` is designed to be used in the browser only. Both `self` and `navigator` are standard properties of HTML5 spec. While it is possible to make `style-loader` work in other environments, it is not something that is supported or guaranteed to work.

**Comment by FlyingDR** [comment]
@ekulabuhov Since the only barrier for use of `style-loader` into other environments is a test for old IE browsers - what stops this test from returning `false` for these non-standard environments which are not old IE by itself?

Updating [this](https://github.com/webpack-contrib/style-loader/blob/master/addStyles.js#L14) line of code with something like
```return window && document && document.all && !window.atob;```

may solve the problem and still be able to detect IE 6-9. Test itself is taken from [Browserhacks](http://browserhacks.com/#hack-e71d8692f65334173fee715c222cb805).

**Comment by ekulabuhov** [comment]
@FlyingDR There's no problem with fixing this line. The problem is that there's no guarantee that the next feature/bug-fix will not break it. Best advice here is to make your environment match the Browser API as much as possible.

**Comment by FlyingDR** [comment]
It is good point, but since there is real use cases when current code breaks builds - I think it is good idea to either update it to be more safe or at least to add some runtime error in development environment or maybe FAQ question so people that will not need to spend much time trying to find out why their build get broken by simple CSS attachment.

**Comment by ekulabuhov** [comment]
I agree. And it should throw a runtime error as implemented here: https://github.com/webpack-contrib/style-loader/commit/0d5089b5f6815cc3a21a9c76e15ebb66cc80d7b0. But it seems like some test environments are able to bypass that check. Meaning that they simulate some parts of the browser API while leaving out other.

**Comment by michael-ciniawsky** [comment]
@FlyingDR Mind sending a PR with eventual fixes and to better discuss the issue?

**Comment by FlyingDR** [comment]
@michael-ciniawsky I'm not quite sure if PR with solution [proposed](https://github.com/webpack-contrib/style-loader/issues/177#issuecomment-284414001) by me above will be acceptable, taking in mind obviously [correct](https://github.com/webpack-contrib/style-loader/issues/177#issuecomment-284421710) [comments](https://github.com/webpack-contrib/style-loader/issues/177#issuecomment-284432423) from @ekulabuhov 

Personally I still think that it is quite safe to check for property existence before accessing them to make assumption about environment, code is running in, but this point may not be accepted by this project, I'm not part of. I can, of course, prepare PR based on my version of the fix.

**Comment by michael-ciniawsky** [comment]
@FlyingDR Well I'm not sure about it 🙃 , seems it happend here https://github.com/webpack-contrib/style-loader/commit/9dc45a6dee16553a64071031d18feb1562df86a2 if you not pissed of enterily if the PR gets rejected, send one please for futher discussion, otherwise I fully understand if you're not interested at all 😛 

> :warning: It's definitely agreed upon that style-loader should only work in browsers enviroments (no SSR-ish]) behaviour

**Comment by FlyingDR** [comment]
@michael-ciniawsky I've added pull request #196

**Comment by d3viant0ne** [comment]
Released and published as `style-loader@0.16.0`

https://github.com/webpack-contrib/style-loader/blob/v0.16.0/CHANGELOG.md#change-log
